src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs

```diff
@@ -0,0 +1,166 @@
+                long wff = OrigWeight(phNodes[winner].First(), phNodes[loser].First());
+                long wfl = OrigWeight(phNodes[winner].First(), phNodes[loser].Last());
+                long wlf = OrigWeight(phNodes[winner].Last(), phNodes[loser].First());
+                long wll = OrigWeight(phNodes[winner].Last(), phNodes[loser].Last());
+                if (wlf >= wff && wlf >= wfl && wlf >= wll)
+                {
+                    // Already in right order
+                }
+                else if (wll >= wff && wll >= wfl && wll >= wlf)
+                {
+                    phNodes[loser].Reverse();
+                }
+                else if (wff >= wfl && wff >= wlf && wff >= wll)
+                {
+                    phNodes[winner].Reverse();
+                }
+                else
+                {
+                    Debug.Assert(wfl >= wff && wfl >= wlf && wfl >= wll);
+                    phNodes[winner].Reverse();
+                    phNodes[loser].Reverse();
+                }
+
+                phNodes[winner].AddRange(phNodes[loser]);
+                phNodes[loser].Clear();
+
+                // Verify that there is exactly one edge between winner's set and loser's set
+                Debug.Assert(phEdges[loser].Count(e => unionFind.FindSet(e.Key) == winner) == 1);
+
+                // Get rid of unifying edge
+                phEdges[winner].Remove(loser);
+                phEdges[loser].Remove(winner);
+
+                // Transfer all edges from loser to winner, coalescing when there are multiple.
+                foreach (var edge in phEdges[loser])
+                {
+                    // Remove counter edge.
+                    bool removed = phEdges[edge.Key].Remove(loser);
+                    Debug.Assert(removed);
+
+                    // Add edge and counter edge, coalescing when necessary.
+                    AddEdge(winner, edge.Key, edge.Value);
+                    AddEdge(edge.Key, winner, edge.Value);
+                    // Add a new entry in the queue as the edge could have changed weight from coalescing.
+                    long weight = phEdges[winner][edge.Key];
+                    queue.Enqueue((winner, edge.Key), -weight); // Priority queue gives lowest priority first
+                }
+
+                phEdges[loser].Clear();
+
+#if DEBUG
+                // Assert that there are only edges between representatives.
+                for (int i = 0; i < phEdges.Length; i++)
+                {
+                    foreach (var edge in phEdges[i])
+                    {
+                        Debug.Assert(unionFind.FindSet(i) == i && unionFind.FindSet(edge.Key) == edge.Key);
+                        Debug.Assert(phEdges[edge.Key][i] == phEdges[i][edge.Key]);
+                    }
+                }
+#endif
+            }
+
+            return
+                phNodes
+                .Where(n => n.Count != 0)
+                .OrderByDescending(n => n.Count)
```

Comment:
Please add a comment here that the stability of this sort is important as it preserves the default sort order for methods which are not described by optimization data as having executed. This is important as the existing default sort order is known to be substantially better than a randomized order.

Reply:
I added a comment and an assert for this. I did not know that this was important (and it worked by accident), so it's good that you said this.